bitcoin/src/bip32.rs

```diff
@@ -256,108 +368,272 @@ impl serde::Serialize for ChildNumber {
+    #[cfg(feature = "serde")]
+    internals::serde_string_impl!(
+        NormalDerivationPath,
+        "a BIP-32 derivation path containing normal numbers only"
+    );
+
+    impl NormalDerivationPath {
+        /// The Master (empty) path (i.e. the master key).
+        #[allow(dead_code)]
+        const MASTER: Self = NormalDerivationPath(Vec::new());
+
+        /// Creates a new [`NormalDerivationPath`] from a vector of [`NormalChildNumber`]s.
+        pub fn new(vec: Vec<NormalChildNumber>) -> Self { NormalDerivationPath(vec) }
+        /// Returns length of the [`NormalDerivationPath`].
+        pub fn len(&self) -> usize { self.0.len() }
+        /// Returns a [`DerivationPath`] from the [`NormalDerivationPath`].
+        pub fn to_derivation_path(&self) -> DerivationPath {
+            self.into_iter().map(|n| n.to_number()).collect()
+        }
 
-#[cfg(feature = "serde")]
-internals::serde_string_impl!(DerivationPath, "a BIP-32 derivation path");
+        /// Returns `true` if the [`NormalDerivationPath`] is empty.
+        pub fn is_empty(&self) -> bool { self.0.is_empty() }
 
-impl<I> Index<I> for DerivationPath
-where
-    Vec<ChildNumber>: Index<I>,
-{
-    type Output = <Vec<ChildNumber> as Index<I>>::Output;
+        /// Returns whether [`NormalDerivationPath`] represents master key (i.e. it's length
+        /// is empty). True for `m` path.
+        pub fn is_master(&self) -> bool { self.is_empty() }
 
-    #[inline]
-    fn index(&self, index: I) -> &Self::Output { &self.0[index] }
-}
+        /// Creates a new [`NormalDerivationPath`] that is a child of this one.
+        pub fn child(&self, cn: NormalChildNumber) -> Self { self.new_extended_from([cn]) }
 
-impl Default for DerivationPath {
-    fn default() -> DerivationPath { DerivationPath::master() }
-}
+        /// Converts into a [`NormalDerivationPath`] that is a child of this one.
+        pub fn into_with_child(self, cn: NormalChildNumber) -> Self {
+            let mut path = self;
+            path.extend_from([cn]);
+            path
+        }
 
-impl<T> IntoDerivationPath for T
-where
-    T: Into<DerivationPath>,
-{
-    fn into_derivation_path(self) -> Result<DerivationPath, Error> { Ok(self.into()) }
-}
+        /// Concatenates `path` to `self`.
+        ///
+        /// ```
+        /// use bitcoin::bip32::{NormalDerivationPath, NormalChildNumber};
+        /// use std::str::FromStr;
+        ///
+        /// let mut base = NormalDerivationPath::from_str("m/42").unwrap();
+        ///
+        /// base.extend_from(NormalDerivationPath::from_str("0/1").unwrap());
+        /// base.extend_from([
+        ///     NormalChildNumber::ZERO,
+        ///     NormalChildNumber::ONE
+        /// ]);
+        ///
+        /// assert_eq!(base, NormalDerivationPath::from_str("m/42/0/1/0/1").unwrap());
+        /// ```
+        pub fn extend_from<T: IntoIterator<Item = NormalChildNumber>>(&mut self, path: T) {
+            self.extend(path);
```

Comment:
What's the point of this if there's already extend?

Reply:
abstraction before calling extend. redundant but i would prefer to maintain this function rather than deleting it and further rewriting it back because i dont want thousand impls on the code

Reply:
Why can't you just replace `extend_from` calls with `extend`?

Reply:
ok